src/rules/noObjectComparisonRule.ts

```diff
@@ -0,0 +1,181 @@
+/**
+ * @license
+ * Copyright 2019 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { isBinaryExpression, isTypeFlagSet } from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+import { codeExamples } from "./code-examples/noObjectComparison.examples";
+
+const OPTION_ALLOW_EQUAL = "allow-equal";
+
+interface Options {
+    allowEquals: boolean;
+}
+
+export class Rule extends Lint.Rules.TypedRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-object-comparison",
+        description: "Only allow comparisons between primitives.",
+        optionsDescription: Lint.Utils.dedent`
+                One argument may be optionally provided:
+                * \`${OPTION_ALLOW_EQUAL}\` allows \`!=\` \`==\` \`!==\` \`===\` comparison between any types.`,
```

Comment:
Although TSLint has traditionally had rule arguments provided as strings, that's starting to be a bit of a pain when they become more complex. This rule seems powerful enough that folks are going to want to add more options later on. Let's have the rule take in an object:

```ts
[true, {
    "allow-equal": true
}]
```